src/native/libs/System.Native/pal_threading.c

```diff
@@ -30,8 +33,12 @@
 
 struct LowLevelMonitor
 {
+#if !defined(TARGET_WASI)
```

Comment:
Didn't check what is the common pattern in pal but I'd expect all such ifdefs to be `#if HAVE_PTHREAD_H`

@jkotas  thoughts?

Reply:
Issue with WASI SDK is that the musl-like headers exist, but they don't have all the symbols.
So I will have to go deeper and detect each function separately. 

Intention of this PR is not to go into this level of details, I just want to make it compile as first step.
If that's too ugly to be merged I would have to change my strategy and skip whole files. 
Please let me know @marek-safar

Reply:
It could be too ugly buy want to hear other opinions too e.g. from Jan

Reply:
This PR is adding a lot of TARGET_WASI ifdefs to System.Native. We did not need these ifdefs for Browser. I assume that Browser Wasm had stubbed out implementations for these. Do you expect that Wasi SDK is going to follow the same plan as Browser eventually? If Wasi SDK does not provide these stubbed out implementations yet, can we provide our own stubbed out implementations instead of adding the ifdefs everywhere?

Reply:
Yes, emscripten SDK has very complete POSIX emulation implemented in JS. For most part it looks like regular unix.

WASI API is not POSIX. WASI-SDK has mapping layer in it.

The goal of WASI is to be expose only those APIs which could be made safe via "capability based security model". Standardization of that is very slow progress. 

Some of the POSIX APIs dotnet use are probably never going to be there. For example terminal control characters, process, groups, users.

The POSIX layer of WASI-SDK is mapping only to the underlying WASI API only when they exist.
It could be possibly switched to full fledged POSIX **headers** via `__wasilibc_unmodified_upstream`, but I think we would have to implement dummy/empty implementation for anything that libs.native uses and linker would be looking for.
It will also defeat the cmake config/detection.

Having our own POSIX layer on top of the underlying WASI API is another option, but I dislike it. I guess the community POSIX layer will be in better shape.

Yet another alternative would be to give up POSIX and have 100% separate libs.native PAL linking directly to WASI APIs.
I have not explored that yet. Downside is that normal development on dotnet would ignore those new files and they will start loosing functional parity.

And finally, I could also exclude most of the PAL `.c` files from the cmake files for now and only include those PAL files which I need and which I fully processed. It would make this PR smaller, but later we would arrive at those same ifdefs and more.

Reply:
Ok, thanks for the context. 

I think it would look best to have the dummy stubs in a separate .c files instead of ifdefing body of nearly every method. It would match the code style that we use for managed components where we preffer separate files too.

For example, pal_threading.c for wasi can look like this:

```
struct LowLevelMonitor
{
   bool dummy
};

LowLevelMonitor* SystemNative_LowLevelMonitor_Create(void)
{
    return (LowLevelMonitor *)malloc(sizeof(LowLevelMonitor));
}

void SystemNative_LowLevelMonitor_Destroy(LowLevelMonitor* monitor)
{
    free(monitor);
}

void SystemNative_LowLevelMonitor_Acquire(LowLevelMonitor* monitor)
{
}

void SystemNative_LowLevelMonitor_Release(LowLevelMonitor* monitor)
{
}

void SystemNative_LowLevelMonitor_Wait(LowLevelMonitor* monitor)
{
}

int32_t SystemNative_LowLevelMonitor_TimedWait(LowLevelMonitor *monitor, int32_t timeoutMilliseconds)
{
    return true;
}

void SystemNative_LowLevelMonitor_Signal_Release(LowLevelMonitor* monitor)
{
}

int32_t SystemNative_CreateThread(uintptr_t stackSize, void *(*startAddress)(void*), void *parameter)
{
    return false;
}
```

Reply:
I created wasi specific files for 
```
pal_console_wasi.c
pal_dynamicload_wasi.c
pal_mount_wasi.c
pal_networking_wasi.c
pal_process_wasi.c
pal_signal_wasi.c
pal_threading_wasi.c
pal_uid_wasi.c
```
they are mostly empty or returning error.

I keep the mixed changes in the `pal_io.c` because part of the functionality is shared.

In some of the next PRs I will probably also get back to shared version of `pal_networking.c` because there is some limited networking in WASI, but for now it's just empty one.